src/Lean/Parser/Module.lean

```diff
@@ -52,14 +52,20 @@ private def mkErrorMessage (c : InputContext) (s : ParserState) (e : Parser.Erro
     e := { e with unexpected }
     -- if there is an unexpected token, include preceding whitespace as well as the expected token could
     -- be inserted at any of these places to fix the error; see tests/lean/1971.lean
-    if let .original (trailing := trailing) .. := s.stxStack.back.getTailInfo then
+    if let some trailing := lastTrailing stk then
       if trailing.stopPos == pos then
         pos := trailing.startPos
   { fileName := c.fileName
     pos := c.fileMap.toPosition pos
     endPos := c.fileMap.toPosition <$> endPos?
     keepFullRange := true
     data := toString e }
+where
+  -- Error recovery might lead to there being some "junk" on the stack
+  lastTrailing (s : SyntaxStack) : Option Substring :=
+    if s.size < 1 then none
+    else if let .original (trailing := trailing) .. := s.back.getTailInfo then some trailing
+    else lastTrailing s.pop
```

Comment:
I think this is `s.toSubarray.findRev? (fun | .original (trailing := trailing) .. => some trailing | _ => none)`?

Reply:
No, because the array contains syntax rather than sourceinfo. It would work to map `getTailInfo` over it first, but it seems wasteful in a strict language.

This does work:
```
    s.toSubarray.findSomeRevM? (m := Id) fun stx =>
      if let .original (trailing := trailing) .. := stx.getTailInfo then pure (some trailing)
        else none
```

Reply:
I just pushed it. I'm not sure that the code is any easier to read, but de gustibus non disputandum est :)